x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/BlobContainerSupplier.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.searchablesnapshots.store;
+
+import org.elasticsearch.common.blobstore.BlobContainer;
+import org.elasticsearch.common.blobstore.OperationPurpose;
+import org.elasticsearch.common.blobstore.support.FilterBlobContainer;
+import org.elasticsearch.repositories.IndexId;
+import org.elasticsearch.repositories.blobstore.BlobStoreRepository;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.function.Supplier;
+
+public class BlobContainerSupplier implements Supplier<BlobContainer> {
+
+    private final Supplier<BlobStoreRepository> repositorySupplier;
+    private final IndexId indexId;
+    private final int shardId;
+
+    private volatile LastKnownState lastKnownState = new LastKnownState(null, null);
+
+    public BlobContainerSupplier(Supplier<BlobStoreRepository> repositorySupplier, IndexId indexId, int shardId) {
+        this.repositorySupplier = repositorySupplier;
+        this.indexId = indexId;
+        this.shardId = shardId;
+    }
+
+    @Override
+    public BlobContainer get() {
+        final LastKnownState lastKnownState = this.lastKnownState;
+        final BlobStoreRepository currentRepository = repositorySupplier.get();
+
+        if (lastKnownState.blobStoreRepository() == currentRepository) {
+            return lastKnownState.blobContainer();
+        } else {
+            return refreshAndGet();
+        }
+    }
+
+    private synchronized BlobContainer refreshAndGet() {
+        final LastKnownState lastKnownState = this.lastKnownState;
```

Comment:
Is this local variable necessary? We are inside synchronized and this is the only place where `this.lastKnownState` can be updated?